# Data Analysis Report: Hospital Y Records Analysis for the Year 2022

Hospital Y has sought assistance in analyzing their records for the year 2022 to make informed decisions. The dataset provided by Hospital Y is in Excel format and consists of four worksheets named 'Visit Tbl', 'Invoice Tbl', 'Diagnosis Tbl', and 'Data Dictionary'. Our objective is to answer the questions provided by the hospital using pandas data frames. 

### Questions to Address

1. How many visits did Kimathi Street and Pipeline medical centers' have from May 2022 and September 2022?

2. What was the most common diagnosis in 2022 for Tassia and Embakasi branches combined?

3.  Which payor was the most profitable (in absolute numbers) for Hospital Y in 2022? {Assume a gross average margin of 30% per visit}.

4. What was the average spend per visit for visits that had a diagnosis of acute gastritis?

5. At Hospital Y, we have a blended healthcare model where patients can get treatment physically (by visiting a medical centers)or virtually (through our call center). In 2022, how many unique patients experienced a blended healthcare approach in their healthcare journey?

6. What was the Net Promoter Score (NPS) for Q3 2022?

7. What was the propotion of the second visits?

8. What percent of visits in April 2022 happened within 30 days of the preceding visit by the same patient?


### Data Analysis Approach
To gain a comprehensive view of the data, we will combine the information from the 'Visit Tbl', 'Invoice Tbl', and 'Diagnosis Tbl' worksheets into a single dataframe. We will exclude the 'Data Dictionary' worksheet as it does not contain relevant data for our analysis. The 'VisitCode' column, present in all three worksheets, will serve as the key to merge the data effectively. By merging the datasets, we can answer our questions based on the integrated information.

Let's proceed with the analysis and present the answers in an organized and clear manner to help Hospital Y gain insights from the data.

In [1]:
# import neccessary modules
import pandas as pd
import numpy as np

In [2]:
# load the excel workbook
workbook = pd.ExcelFile("D:/MyProjects/BI_Analyst_Assessment_Data_2023.xlsx")

# get the worksheet names
workbook.sheet_names

['Visit Tbl', 'Invoice Tbl', 'Diagnosis Tbl', 'Data Dictionary']

In [3]:
# read the worksheets into a DataFrame
sheet_1 = pd.read_excel("BI_Analyst_Assessment_Data_2023.xlsx",sheet_name = "Visit Tbl")
sheet_2 = pd.read_excel("BI_Analyst_Assessment_Data_2023.xlsx",sheet_name = "Invoice Tbl")
sheet_3 = pd.read_excel("BI_Analyst_Assessment_Data_2023.xlsx",sheet_name = "Diagnosis Tbl")

# merge the sheets on "VisitCode" column
df = pd.merge(sheet_1,sheet_2,on = "VisitCode")
df = pd.merge(df,sheet_3,on ="VisitCode")

# first five records after merging
df.head()

VisitCode    PatientCode       VisitDateTime MedicalCenter  \
0  XA-1188078  c464fa34-e7f2 2022-07-27 05:54:02        Tassia   
1  XA-1220192  7b8216fe-ff14 2022-09-06 12:28:24        Tassia   
2  XA-1120632  ae928861-7de6 2022-05-17 20:37:25        Tassia   
3  XA-1146859  758bee68-b569 2022-06-12 16:41:30        Tassia   
4  XA-1204110  1c570cd7-0b55 2022-08-16 13:34:48        Tassia   

     VisitCategory Payor  NPS Score  Amount                     Diagnosis  
0  In-person Visit  Cash        NaN    1800                        review  
1  In-person Visit  Cash        NaN    1460                acute rhinitis  
2  In-person Visit  Cash        NaN     300                        review  
3  In-person Visit  Cash        NaN    1515  tonsillitis, acute bacterial  
4  In-person Visit  Cash        8.0    4140                     pneumonia

In [4]:
# dataframe info after merging
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22119 entries, 0 to 22118
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   VisitCode      22119 non-null  object        
 1   PatientCode    22119 non-null  object        
 2   VisitDateTime  22119 non-null  datetime64[ns]
 3   MedicalCenter  22119 non-null  object        
 4   VisitCategory  22119 non-null  object        
 5   Payor          22119 non-null  object        
 6   NPS Score      852 non-null    float64       
 7   Amount         22119 non-null  int64         
 8   Diagnosis      22119 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(6)
memory usage: 1.7+ MB


From the DataFrame information, it appears that the "NPS Score" column contains 852 null values out of a total of 22,119 entries. While this represents a significant number of missing values, we have decided not to drop the column as it is essential for answering the questions posed by Hospital Y. Instead, we will handle these null values appropriately during our analysis to ensure that they do not affect the validity of our results.

In [5]:
# check for duplicates
df.duplicated().value_counts()

False    21863
True       256
dtype: int64

In [6]:
# drop duplicates
df = df.drop_duplicates()
df.head()

VisitCode    PatientCode       VisitDateTime MedicalCenter  \
0  XA-1188078  c464fa34-e7f2 2022-07-27 05:54:02        Tassia   
1  XA-1220192  7b8216fe-ff14 2022-09-06 12:28:24        Tassia   
2  XA-1120632  ae928861-7de6 2022-05-17 20:37:25        Tassia   
3  XA-1146859  758bee68-b569 2022-06-12 16:41:30        Tassia   
4  XA-1204110  1c570cd7-0b55 2022-08-16 13:34:48        Tassia   

     VisitCategory Payor  NPS Score  Amount                     Diagnosis  
0  In-person Visit  Cash        NaN    1800                        review  
1  In-person Visit  Cash        NaN    1460                acute rhinitis  
2  In-person Visit  Cash        NaN     300                        review  
3  In-person Visit  Cash        NaN    1515  tonsillitis, acute bacterial  
4  In-person Visit  Cash        8.0    4140                     pneumonia

In [7]:
# dataframe information after cleaning our data
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21863 entries, 0 to 22118
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   VisitCode      21863 non-null  object        
 1   PatientCode    21863 non-null  object        
 2   VisitDateTime  21863 non-null  datetime64[ns]
 3   MedicalCenter  21863 non-null  object        
 4   VisitCategory  21863 non-null  object        
 5   Payor          21863 non-null  object        
 6   NPS Score      845 non-null    float64       
 7   Amount         21863 non-null  int64         
 8   Diagnosis      21863 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(6)
memory usage: 1.7+ MB


Everything looks good now!Lets dive into our analysis

In [8]:
# medical centers in Penda health
df["MedicalCenter"].unique()

array(['Tassia', 'Pipeline', 'Embakasi', 'Githurai 45', 'Mathare North',
       'Kimathi Street', 'Lucky Summer', 'Call Centre'], dtype=object)

### How many visits did Kimathi Street and Pipeline medical centers' have from May 2022 and September 2022?

In [9]:
visit_df = df[(df["MedicalCenter"] == "Kimathi Street") | (df["MedicalCenter"] == "Pipeline")]

len(visit_df[visit_df["VisitDateTime"].between("2022-05-01","2022-09-30")])

4626

### What was the most common diagnosis in 2022 for Tassia and Embakasi branches combined?

In [10]:
# five most common diagnosis in all medical centers
diagnosis_medicalcenter = df.groupby(["MedicalCenter","Diagnosis"])["Diagnosis"].count().sort_values(ascending=False)
diagnosis_medicalcenter.head()

MedicalCenter  Diagnosis                   
Pipeline       acute nasopharyngitis           1640
Tassia         acute nasopharyngitis           1088
               tonsillitis, acute bacterial     936
Pipeline       tonsillitis, acute bacterial     735
               acute gastritis                  636
Name: Diagnosis, dtype: int64

In [11]:
# tassia
df_grouped = df.groupby(["MedicalCenter", "Diagnosis"]).size().reset_index(name='Count')
most_common_diagnosis_Tassia = df_grouped[df_grouped["MedicalCenter"] == "Tassia"].sort_values(by = "Count",ascending=False)

print(f"The most common diagnosis in Tassia Medical Center was {most_common_diagnosis_Tassia.Diagnosis.iloc[0]}")

The most common diagnosis in Tassia Medical Center was acute nasopharyngitis


In [12]:
# embakasi
most_common_diagnosis_Embakasi =  df_grouped[df_grouped["MedicalCenter"] == "Embakasi"].sort_values(by = "Count",ascending=False)

print(f"The most common diagnosis in Embakasi Medical Center was {most_common_diagnosis_Embakasi.Diagnosis.iloc[0]}")

The most common diagnosis in Embakasi Medical Center was acute nasopharyngitis


### Which payor was the most profitable (in absolute numbers) for Hospital Y in 2022? {Assume a gross average margin of 30% per visit}.

In [13]:
# calculate profit per medical center
df["Revenue"] = df["Amount"].apply(lambda x:x*0.30)
df.head()

VisitCode    PatientCode       VisitDateTime MedicalCenter  \
0  XA-1188078  c464fa34-e7f2 2022-07-27 05:54:02        Tassia   
1  XA-1220192  7b8216fe-ff14 2022-09-06 12:28:24        Tassia   
2  XA-1120632  ae928861-7de6 2022-05-17 20:37:25        Tassia   
3  XA-1146859  758bee68-b569 2022-06-12 16:41:30        Tassia   
4  XA-1204110  1c570cd7-0b55 2022-08-16 13:34:48        Tassia   

     VisitCategory Payor  NPS Score  Amount                     Diagnosis  \
0  In-person Visit  Cash        NaN    1800                        review   
1  In-person Visit  Cash        NaN    1460                acute rhinitis   
2  In-person Visit  Cash        NaN     300                        review   
3  In-person Visit  Cash        NaN    1515  tonsillitis, acute bacterial   
4  In-person Visit  Cash        8.0    4140                     pneumonia   

   Revenue  
0    540.0  
1    438.0  
2     90.0  
3    454.5  
4   1242.0

In [14]:
# groupby payor
payor_df = df.groupby("Payor")["Revenue"].sum().sort_values(ascending=False)

print(f"The most profitable payor for Hospital Y in 2022 was {payor_df.index[0]}")

The most profitable payor for Hospital Y in 2022 was Insurance Company B


### What was the average spend per visit for visits that had a diagnosis of acute gastritis?

In [15]:
# select rows with diagnosis "acute gastritis"
acute_gastritis = df[df["Diagnosis"] == "acute gastritis"]

# find the mean of the amount
acute_gastritis.Amount.mean()

print(f"The average spend per visit that had a diagnosis of acute gastritis was {acute_gastritis.Amount.mean():.2f}")

The average spend per visit that had a diagnosis of acute gastritis was 2777.57


### At Penda Health, we have a blended healthcare model where patients can get treatment physically (by visiting a medical centers)or virtually (through our call center). In 2022, how many unique patients experienced a blended healthcare approach in their healthcare journey?

In [16]:
df.groupby("VisitCategory")["PatientCode"].count()

VisitCategory
In-person Visit       21184
Telemedicine Visit      679
Name: PatientCode, dtype: int64

### What was the Net Promoter Score (NPS) for Q3 2022?

Since we want to retain the NPS Score column in the DataFrame and handle the missing values (NaN), we will fill the NaN values with the mode of the NPS Score column. This approach allows us to replace the missing values with the most frequently occurring value in the column, ensuring that the data remains representative and complete.

In [17]:
# fill the null values with the mode
mode = df["NPS Score"].mode().iloc[0]
df["NPS Score"].fillna(mode,inplace = True)

In [18]:
# get_review function to classify the nps scores
def get_review(score):
    if 9.0 <= score <= 10.0:
        return "Promoters"
    elif 7.0 <= score <= 10.0:
        return "Passive"
    else:
        return "Detractors"

In [19]:
# apply the function to nps score column and create a new column for the scores
df["Reviews"] = df["NPS Score"].apply(get_review)

In [20]:
# calculate the NPS score
percentage_promoters = len(df[(df["Reviews"] == "Promoters") | (df["VisitDateTime"].between("2022-07-01", "2022-09-30"))]) / len(df["Reviews"]) * 100

percentage_detractors = len(df[(df["Reviews"] == "Detractors") | (df["VisitDateTime"].between("2022-07-01", "2022-09-30"))]) / len(df["Reviews"]) * 100

NPS = percentage_promoters - percentage_detractors
print(f"The NPS Score for Penda Health was {NPS}. This shows that the patients were satisfied with the services and would recommend other customers")


The NPS Score for Penda Health was 65.10085532635046. This shows that the patients were satisfied with the services and would recommend other customers


### What was the propotion of the second visits?

In [21]:
# group the visitcode with unique patient code
second_visits = df.groupby("VisitCode")["PatientCode"].count()

# propotion of second visits 
propotion_of_second_visits = len(second_visits.loc[second_visits == 2])/len(second_visits)
propotion_of_second_visits

# print the proportion of second visits
print(f"The proportion of second visits in Penda Health in 2022 is: {propotion_of_second_visits:.2f}")

The proportion of second visits in Penda Health in 2022 is: 0.10


### What percent of visits in April 2022 happened within 30 days of the preceding visit by the same patient?

In [22]:
# filter the visits in April 2022
visits_april_2022 = df[(df['VisitDateTime'].dt.year == 2022) & (df['VisitDateTime'].dt.month == 4)]

# sort the visits by patient ID and visit date
sorted_visits = visits_april_2022.sort_values(['PatientCode', 'VisitDateTime'])

# calculate the time difference between consecutive visits for each patient
sorted_visits['TimeDifference'] = sorted_visits.groupby('PatientCode')['VisitDateTime'].diff()

# count the number of visits that occurred within 30 days of the preceding visit
within_30_days = len(sorted_visits[sorted_visits['TimeDifference'] <= pd.Timedelta(days=30)])

# calculate the percentage
total_visits = len(visits_april_2022)
percentage_within_30_days = (within_30_days / total_visits) * 100

# print the percentage
print(f"The percentage of visits in April 2022 within 30 days of the preceding visit is {percentage_within_30_days:.2f}%")


The percentage of visits in April 2022 within 30 days of the preceding visit is 16.24%
